pkg/virt-controller/watch/testing/testing.go

```diff
@@ -101,3 +104,28 @@ func DefaultVirtualMachineWithNames(started bool, vmName string, vmiName string)
 func DefaultVirtualMachine(started bool) (*v1.VirtualMachine, *v1.VirtualMachineInstance) {
 	return DefaultVirtualMachineWithNames(started, "testvmi", "testvmi")
 }
+
+type interf interface {
+	Execute() bool
+}
+
+func deepCopyList(objects []interface{}) []interface{} {
+	for i := range objects {
+		objects[i] = objects[i].(runtime.Object).DeepCopyObject()
+	}
+	return objects
+}
+
+func SanityExecute(c interf, stores []cache.Store, g gomega.Gomega) {
```

Comment:
I have a proposal, why don't we use reflection here to automatically populate the store list with all the stores present in the `Controller` struct? I have written this version here which I tested and works

```go
func SanityExecute(c interf, g gomega.Gomega) {
	listOfObjects := [][]interface{}{}

	storeType := reflect.TypeOf((*cache.Store)(nil)).Elem()
	ctrlType := reflect.Indirect(reflect.ValueOf(c))

	var stores []cache.Store
	for i := range ctrlType.NumField() {
		field := ctrlType.Field(i)
		if field.Type().Implements(storeType) {
			storeField := reflect.NewAt(field.Type(), unsafe.Pointer(field.UnsafeAddr())).Elem()
			stores = append(stores, storeField.Interface().(cache.Store))
		}
	}

	for _, store := range stores {
		listOfObjects = append(listOfObjects, deepCopyList(store.List()))
	}

	c.Execute()

	for i, objects := range listOfObjects {
		g.ExpectWithOffset(1, stores[i].List()).To(gomega.ConsistOf(objects...))
	}
}
```

This way you can just call `SanityExecute` this way

```go
sanityExecute := func() {
    watchtesting.SanityExecute(controller, Default)
}
```

and it becomes generic for all controllers.

On top of that it'd be great if you create a test controller wrapper using generics to automatically call sanityExecute.

Reply:
Btw, reflection is not good enough, you would need to nest into fields as they can also contain `Store`s

Reply:
@xpivarc that is true, but wow that would be weird... I wouldn't approve a PR that introduces that.

You can anyway to a DFS into the fields and make sure you have everything, anyway I think this looks good for now but I wouldn't give up into something more complex but transparent to the developer.

Reply:
fiy, we do have nested `Store` :/